keps/2936-local-queue-defaulting/README.md

```diff
@@ -0,0 +1,63 @@
+# KEP-2936: LocalQueue defaulting
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces a defaulting LocalQueue mechanism for the jobs that don't have
+LocalQueue name annotation specified.
+
+## Motivation
+
+Simplify user experience with Kueue.
+
+### Goals
+
+- default LocalQueue for jobs that don't specify LocalQueue name annotation.
+
+### Non-Goals
+
+- create a default LocalQueue.
+
+## Proposal
+
+If the Job doesn't have annotation for LocalQueue in its Spec and the LocalQueue
+with name `default` is present in Job's namespace and KueueDefaulting feature gate
+is enabled, the `kueue.x-k8s.io/queue-name:default` anotation will be added to the Job.
+
+### User Stories
+
+#### Story 1
+
+Administrator createds a single `default` LocalQueue per namespace. Users doesn't need to specify
+the LocalQueue annotation in order for workload to be scheduled.
+
+### Notes/Constraints/Caveats (Optional)
+
+manageJobsWithoutQueueName doesn't work with this feature, since Kueue add the LocalQueue
+annotation.
+
+### Risks and Mitigations
+
+## Design Details
+
+Update defautling webhook for all types of job to add `kueue.x-k8s.io/queue-name:default`
```

Comment:
Could you expand on the proposed approach? The options I see:
1. default the `queue-name` label on the Job object in a webhook
2. default only at the workload `spec.queueName` level

Since we cannot control the order to webhooks I think for (1.) we would need to be the Job's webhook. However, this would not work OOTB for external Jobs. Also, with the MAP effort we will want to move away from webhooks over time. On the plus side we have compatibility with tooling.

(2.) would work OOTB for external Jobs. Also, (2.) is less committal, we could always do (1.) in a follow up if needed. 

Maybe the issue for (2.) is the compatibility with other tooling like kueuectl (not sure where it looks now), but we could probably adjust them to lookup the workload.

All in all, I'm not sure which one is better, so would be good to put list the pros and cons, and put the rejected one to alternatives describing the reasons for rejecting.

Reply:
>The webhook-based approach will not work OOTB for external Jobs.

Could you elaborate why it won't work?

>Also, with the MAP effort we will want to move away from webhooks over time.

What is MAP?

Reply:
> Could you elaborate why it won't work?

It will work, but not OOTB - the users need to implement the same semantics in their webhooks. AppWrapper is one of the frameworks, but there are also other in-house Jobs.

> What is MAP?

[Mutating Admission Policies](https://github.com/kubernetes/enhancements/issues/3962). I think we could actually move the approach from webhook to MAP when it is released.

Reply:
>Actually, (2.) may not work OOTB either - because if the webhook is not aware of the logic, then it would suspend the job without queue-name when manageJobsWithoutName: true (IIUC)

yes, this is my understanding

Reply:
I seem in that case I'm leaning to (1.) as IIUC both approaches don't work OOTB for external Jobs. It would be great to make it easy for developers of custom Jobs to easily support the semantics, which becomes non-trivial, as the webhooks needs to know if the LQ exists (so needs access to cache or have LQ informer).

Reply:
It's going to be important that the webhook only do this for Jobs defined in namespaces where a designated default `LocalQueue` already exists.  If the webhook acts  without namespace awareness, then we still need the namespace-level filtering mechanism from[ KEP-3589](https://github.com/kubernetes-sigs/kueue/pull/3595)

Reply:
I think the webhook should be namespace aware, but the awareness would come from  KEP-3589, so that we don't duplicate namespace filtering effort.

Reply:
I check whether the `default`queue is exists in the namespace: https://github.com/kubernetes-sigs/kueue/pull/3610

Reply:
> If we go with (1.) we should provide some utility functions so that it is easy for other developers to follow the semantics for the external jobs.

I think that this should be enough for the in-house jobs. Additionally, we might want to expand the customJob documentation (https://kueue.sigs.k8s.io/docs/tasks/dev/integrate_a_custom_job/).

Basically, I think that natively supporting custom Jobs is not necessary in here's upstream repository, and we can focus on the upstream supported Jobs like batch/v1 Job, RayJob, and KFJobs.